src/expr/impl/src/aggregate/approx_percentile.rs

```diff
@@ -12,56 +12,163 @@
 impl AggregateFunction for ApproxPercentile {
     fn return_type(&self) -> DataType {
         DataType::Float64
     }
 
     fn create_state(&self) -> Result<AggregateState> {
-        todo!()
+        Ok(AggregateState::Any(Box::<State>::default()))
     }
 
-    async fn update(&self, _state: &mut AggregateState, _input: &StreamChunk) -> Result<()> {
-        todo!()
+    async fn update(&self, state: &mut AggregateState, input: &StreamChunk) -> Result<()> {
+        let state: &mut State = state.downcast_mut();
+        for (op, row) in input.rows() {
+            let datum = row.datum_at(0);
+            self.add_datum(state, op, datum);
+        }
+        Ok(())
     }
 
     async fn update_range(
         &self,
-        _state: &mut AggregateState,
-        _input: &StreamChunk,
-        _range: Range<usize>,
+        state: &mut AggregateState,
+        input: &StreamChunk,
+        range: Range<usize>,
     ) -> Result<()> {
-        todo!()
+        let state = state.downcast_mut();
+        for (op, row) in input.rows_in(range) {
+            self.add_datum(state, op, row.datum_at(0));
+        }
+        Ok(())
+    }
+
+    // TODO(kwannoel): Instead of iterating over all buckets, we can maintain the
+    // approximate quantile bucket on the fly.
+    async fn get_result(&self, state: &AggregateState) -> Result<Datum> {
+        let state = state.downcast_ref::<State>();
+        let quantile_count = (state.count as f64 * self.quantile) as u64;
+        let mut acc_count = 0;
+        for (bucket_id, count) in &state.buckets {
+            acc_count += count;
+            if acc_count >= quantile_count {
+                // approx value = 2 * y^i / (y + 1)
+                let approx_percentile = 2.0 * self.base.powi(*bucket_id) / (self.base + 1.0);
+                let approx_percentile = ScalarImpl::Float64(approx_percentile.into());
+                return Ok(Datum::from(approx_percentile));
+            }
+        }
+        return Ok(None);
+    }
+
+    fn encode_state(&self, state: &AggregateState) -> Result<Datum> {
+        let state = state.downcast_ref::<State>();
+        let mut encoded_state = Vec::with_capacity(state.estimated_heap_size());
+        encoded_state.extend_from_slice(&state.count.to_be_bytes());
+        for (bucket_id, count) in &state.buckets {
+            encoded_state.extend_from_slice(&bucket_id.to_be_bytes());
+            encoded_state.extend_from_slice(&count.to_be_bytes());
+        }
+        let encoded_scalar = ScalarImpl::Bytea(encoded_state.into());
+        Ok(Datum::from(encoded_scalar))
     }
 
-    async fn get_result(&self, _state: &AggregateState) -> Result<Datum> {
-        todo!()
+    fn decode_state(&self, datum: Datum) -> Result<AggregateState> {
+        let mut state = State::default();
+        let Some(scalar_state) = datum else {
+            return Ok(AggregateState::Any(Box::new(state)));
+        };
+        println!("scalar state: {:?}", scalar_state);
```

Comment:
There's an issue here:
```
scalar state: Float64(OrderedFloat(1.1897e-320))
thread 'rw-streaming' panicked at src/common/src/types/mod.rs:710:1:
cannot convert ScalarImpl::Float64 to concrete type Bytea
```

I get this panic, which doesn't really make sense to me.

In `encode_state` we clearly encode it as a `Bytea`. How come we receive it as a `Float64` here?

Is it something wrong in `infer_state`?

cc @wangrunji0408 @stdrc

Reply:
Nevermind, looks like we need to manually specify the decode state type in the type sig.

Reply:
```
#[build_aggregate("approx_percentile(float8) -> float8", state = "bytea")]
```